glusterd2/commands/volumes/volume-expand.go

```diff
@@ -87,18 +97,55 @@ func volumeExpandHandler(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	for index := range volinfo.Subvols {
-		for _, brick := range volinfo.Subvols[index].Bricks {
+	var expansionSizePerBrick uint64
+	var expansionTpSizePerBrick uint64
+	var expansionMetadataSizePerBrick uint64
+	var brickVgMapping map[string]string
+	var ok bool
+	lvmResizeOp := checkForLvmResize(req, volinfo)
+	// continue normal volume expand by adding new bricks or subvols
+	if !lvmResizeOp {
+		for index := range volinfo.Subvols {
+			for _, brick := range volinfo.Subvols[index].Bricks {
 
-			for _, b := range req.Bricks {
+				for _, b := range req.Bricks {
 
-				if brick.PeerID.String() == b.PeerID && brick.Path == filepath.Clean(b.Path) {
-					restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrDuplicateBrickPath)
-					return
+					if brick.PeerID.String() == b.PeerID && brick.Path == filepath.Clean(b.Path) {
+						restutils.SendHTTPError(ctx, w, http.StatusBadRequest, errors.ErrDuplicateBrickPath)
+						return
+					}
 				}
+
 			}
+		}
+	} else {
+		// lvmResize is needed
+		switch volinfo.Type {
+		case volume.Distribute:
+			expansionSizePerSubvol := req.Size / uint64(len(volinfo.Subvols))
+			expansionSizePerBrick = expansionSizePerSubvol
+		case volume.Replicate, volume.DistReplicate:
+			expansionSizePerSubvol := req.Size / uint64(len(volinfo.Subvols))
+			expansionSizePerBrick = expansionSizePerSubvol
+		case volume.Disperse, volume.DistDisperse:
+			expansionSizePerSubvol := req.Size / uint64(len(volinfo.Subvols))
+			expansionSizePerBrick = expansionSizePerSubvol / uint64(len(volinfo.Subvols[0].Bricks))
```

Comment:
This should be `brickExpandSize = subvolSize / number of data bricks`

```
expansionSizePerBrick = expansionSizePerSubvol / uint64(volinfo.Subvols[0].DisperseCount - volinfo.Subvols[0].RedundancyCount)
```

Reply:
Done